### PR TITLE
Add script to report rules

### PR DESCRIPTION
**Pull request by AyanSinhaMahapatra**
Partially adresses #2566

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

**Comment by AyanSinhaMahapatra** [comment]
@mjherzog this is ready for you to look at.

Here's the help text for this script:

```
Usage: report_license_rules.py [OPTIONS]

  Write Licenses/Rules from scancode into a CSV file with all details. Output
  can be optionally filtered by category/license-key.

Options:
  -l, --licenses FILE       Write all Licenses data to the csv FILE.
  -r, --rules FILE          Write all Rules data to the csv FILE.
  -c, --category STRING     An optional filter to only output licenses/rules
                            of this category. Example STRING: `permissive`.
  -k, --license-key STRING  An optional filter to only output licenses/rules
                            which has this license key.Example STRING: `mit`.
  -t, --with-text           Also include the license/rules texts.Note that
                            this increases the file size significantly.
  -h, --help                Show this message and exit.

```

You have to checkout this branch in a configured virtualenv and run the script if you want to check this out now. 

Attaching sample ouput files:

[licence_data.csv](https://github.com/nexB/scancode-toolkit/files/7107763/lic_data.csv)
[rule_data.csv](https://github.com/nexB/scancode-toolkit/files/7107762/rule_data.csv)

**Comment by mjherzog** [comment]
@AyanSinhaMahapatra Thank you for running the sample with:
python ./etc/scripts/licenses/report_license_rules.py -l comm_licenses_data.csv -r comm_rule_data.csv -t -c "Commercial"
It looks good overall, but there is strange data in the identifier column - it looks like some kind of "bleed" from license text. Easy to see the problem if you sort by identifier because there is a very big chunk at the beginning and the other examples interspersed among the "true" identifiers.

**Comment by mjherzog** [comment]
I see clues for Accellera and Hacktivismo in the errant text.
If we have text too large for a single csv/Excel cell we could just truncate it. This tool is needed most for licenses with relatively sparse text for most rules like commercial-license.

File: etc/scripts/licenses/report_license_rules.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import click
+import csv
+
+from commoncode.cliutils import PluggableCommandLineOption
+from licensedcode.models import load_licenses
+from licensedcode.models import load_rules
+
+
+LICENSES_FIELDNAMES = [
+    'key', 'short_name', 'name', 'category', 'owner',  'text', 'words_count', 'notes',
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py:22]: text is always going to be problematic... using the first `[:200]` characters is going to be enough

File: etc/scripts/licenses/report_license_rules.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    'minimum_coverage', 'homepage_url', 'is_exception', 'language', 'is_unknown',
+    'spdx_license_key', 'text_urls', 'other_urls', 'standard_notice',
+    'license_filename', 'faq_url', 'ignorable_authors',
+    'ignorable_copyrights', 'ignorable_holders', 'ignorable_urls', 'ignorable_emails',
+    'osi_license_key', 'osi_url', 'other_spdx_license_keys',
+]
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py:28]: Add the licensedb URL too

File: etc/scripts/licenses/report_license_rules.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    'is_false_positive', 'is_license_text', 'is_license_notice',  'is_license_tag',
+    'is_license_reference', 'is_license_intro', 'has_unknown', 'only_known_words',
+    'notes', 'referenced_filenames', 'minimum_coverage', 'ignorable_copyrights',
+    'ignorable_holders', 'ignorable_authors', 'ignorable_urls', 'ignorable_emails',   
+]
+
+
+def write_data_to_csv(data, output_csv, fieldnames):
+    
+    with open(output_csv,'w',encoding='utf-8-sig',newline='') as f:
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py]: `,encoding='utf-8-sig',newline=''` are suspicious to me... IMHO you are best of without them

File: etc/scripts/licenses/report_license_rules.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        w = csv.DictWriter(f,fieldnames=fieldnames)
+        w.writeheader()
+
+        for entry in data:
+            w.writerow(entry)
+
+
+def filter_by_attribute(data, attribute, expected):
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py]: may be required_key is a better name than expected?

File: etc/scripts/licenses/report_license_rules.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    
+    filtered_output = []
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py]: my be this is readable too?

```Python
return [entry for entry in data if entry.get(attribute) == expected]
```

File: etc/scripts/licenses/report_license_rules.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    for entry in data:
+        if entry.get(attribute, 'None') == expected:
+            filtered_output.append(entry)
+            
+    return filtered_output
+
+def flatten_output(data):
+
+    if not isinstance(data, list):
+        return
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py]: say I have a tuple or a dict?... the flatten_output is then empty? The implicit exclusion of non-list is a bit surprising. If you only ever expect a list, then instead use `assert isinstance(data, list)`

File: etc/scripts/licenses/report_license_rules.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+    output = []
+    for entry in data:
+        if not isinstance(entry, dict):
```
Comment (pombredanne) [inline etc/scripts/licenses/report_license_rules.py]: You should assert instead

**Review by pombredanne — APPROVED** [review]
See a few nits for your consideration!
Feel free to merge.

**Comment by AyanSinhaMahapatra** [comment]
@pombredanne added requested changes.

@mjherzog added some changes that might fix your problems earlier. Made texts have only first 200 characters and removed some probably problematic csv file handling. Adding new csv file reports, do check if the bugs are resolved.

The commands used to generate reports:
```
python ./etc/scripts/licenses/report_license_rules.py -l licenses_data.csv -r rule_data.csv -t
python ./etc/scripts/licenses/report_license_rules.py -l comm_licenses_data.csv -r comm_rule_data.csv -t -c "Commercial"
```

Reports:
[licenses_data.csv](https://github.com/nexB/scancode-toolkit/files/7115780/licenses_data.csv)
[rule_data.csv](https://github.com/nexB/scancode-toolkit/files/7115782/rule_data.csv)
[comm_licenses_data.csv](https://github.com/nexB/scancode-toolkit/files/7115785/comm_licenses_data.csv)
[comm_rule_data.csv](https://github.com/nexB/scancode-toolkit/files/7115789/comm_rule_data.csv)

**Pull request merged**